src/symbolic_memory.ml

```diff
@@ -111,51 +133,81 @@ module M = struct
+  let calculate_address m (a : int32) : (Int32.t, Trap.t) Stdlib.Result.t =
+    match a.e with
+    | Val (Num (I32 i)) -> Ok i
+    | Ptr (base, offset) -> (
+      match Hashtbl.find_opt m.chunks base with
+      | None -> Error Trap.Memory_leak_use_after_free
+      | Some size ->
+        let ptr = Int32.add base (i32 offset) in
+        if ptr < base || ptr > Int32.add base (i32 size) then
+          Error Trap.Memory_heap_buffer_overflow
+        else Ok ptr )
+    | _ -> Format.kasprintf failwith "Cannot calculate address of: %a" Expr.pp a
+
   let load_8_s m a =
-    let v = loadn m (concretize_i32 a.e) 1 in
+    let* a = calculate_address m a in
+    let v = loadn m a 1 in
     match v.e with
-    | Val (Num (I32 i8)) -> Value.const_i32 (Int32.extend_s 8 i8)
-    | _ -> Cvtop (ExtS 24, v) @: Ty_bitv S32
+    | Val (Num (I8 i8)) ->
+      Ok (Value.const_i32 (Int32.extend_s 8 (Int32.of_int i8)))
+    | _ -> Ok (Cvtop (ExtS 24, v) @: Ty_bitv S32)
 
   let load_8_u m a =
-    let v = loadn m (concretize_i32 a.e) 1 in
+    let* a = calculate_address m a in
+    let v = loadn m a 1 in
     match v.e with
-    | Val (Num (I32 _)) -> v
-    | _ -> Cvtop (ExtU 24, v) @: Ty_bitv S32
+    | Val (Num (I8 i)) -> Ok (Value.const_i32 (Int32.of_int i))
+    | _ -> Ok (Cvtop (ExtU 24, v) @: Ty_bitv S32)
 
   let load_16_s m a =
-    let v = loadn m (concretize_i32 a.e) 2 in
+    let* a = calculate_address m a in
+    let v = loadn m a 2 in
     match v.e with
-    | Val (Num (I32 i16)) -> Value.const_i32 (Int32.extend_s 16 i16)
-    | _ -> Cvtop (ExtS 16, v) @: Ty_bitv S32
+    | Val (Num (I32 i16)) -> Ok (Value.const_i32 (Int32.extend_s 16 i16))
+    | _ -> Ok (Cvtop (ExtS 16, v) @: Ty_bitv S32)
 
   let load_16_u m a =
-    let v = loadn m (concretize_i32 a.e) 2 in
+    let* a = calculate_address m a in
+    let v = loadn m a 2 in
     match v.e with
-    | Val (Num (I32 _)) -> v
-    | _ -> Cvtop (ExtU 16, v) @: Ty_bitv S32
+    | Val (Num (I32 _)) -> Ok v
+    | _ -> Ok (Cvtop (ExtU 16, v) @: Ty_bitv S32)
 
-  let load_32 m a = loadn m (concretize_i32 a.e) 4
+  let load_32 m a =
+    let* a = calculate_address m a in
+    Ok (loadn m a 4)
 
-  let load_64 m a = loadn m (concretize_i32 a.e) 8
+  let load_64 m a =
+    let* a = calculate_address m a in
+    Ok (loadn m a 8)
 
   let extract v pos =
     match v.e with
     | Val (Num (I32 i)) ->
-      let i' = Int32.(logand 0xffl @@ shr_s i @@ of_int (pos * 8)) in
-      Value.const_i32 i'
+      let i' = Int32.(to_int @@ logand 0xffl @@ shr_s i @@ of_int (pos * 8)) in
+      Val (Num (I8 i')) @: Ty_bitv S8
     | Val (Num (I64 i)) ->
-      let i' = Int64.(logand 0xffL @@ shr_s i @@ of_int (pos * 8)) in
-      Value.const_i32 (Int32.of_int64 i')
+      let i' = Int64.(to_int @@ logand 0xffL @@ shr_s i @@ of_int (pos * 8)) in
+      Val (Num (I8 i')) @: Ty_bitv S8
+    | Cvtop (ExtU 24, { e = Symbol s; ty })
+    | Cvtop (ExtS 24, { e = Symbol s; ty })
+      when ty = Ty_bitv S8 ->
```

Comment:
I believe a `when` is not needed here and you can directly put it in the pattern.